File: LLM-Translations/GPT-4/avatar/Python/Go/atcoder_ABC165_C.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func solve(N int, M int, Q int, a []int, b []int, c []int, d []int) {
	res := 0
	var rec func(A []int)
	rec = func(A []int) {
		if len(A) == N {
			ans := 0
			for i := 0; i < Q; i++ {
				if A[b[i]-1]-A[a[i]-1] == c[i] {
					ans += d[i]
				}
			}
			if ans > res {
				res = ans
			}
			return
		}
		last_max := 1
		if len(A) != 0 {
			last_max = A[len(A)-1]
		}
		for i := last_max; i <= M; i++ {
			rec(append(A, i))
		}
	}
	rec([]int{})
	fmt.Println(res)
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Split(bufio.ScanWords)
	N, _ := strconv.Atoi(next(scanner))
	M, _ := strconv.Atoi(next(scanner))
	Q, _ := strconv.Atoi(next(scanner))
	a, b, c, d := make([]int, Q), make([]int, Q), make([]int, Q), make([]int, Q)
	for i := 0; i < Q; i++ {
		a[i], _ = strconv.Atoi(next(scanner))
		b[i], _ = strconv.Atoi(next(scanner))
		c[i], _ = strconv.Atoi(next(scanner))
		d[i], _ = strconv.Atoi(next(scanner))
	}
	solve(N, M, Q, a, b, c, d)
}

func next(scanner *bufio.Scanner) string {
	scanner.Scan()
	return scanner.Text()
}

// End of Code

File: LLM-Translations/GPT-4/avatar/Python/C/atcoder_ABC165_C.c
#include <stdio.h>
#include <stdlib.h>

void rec(int N, int M, int Q, int *a, int *b, int *c, int *d, int *A, int *res) {
    if (N == 0) {
        int ans = 0;
        for (int i = 0; i < Q; i++) {
            if (A[b[i] - 1] - A[a[i] - 1] == c[i]) {
                ans += d[i];
            }
        }
        if (ans > *res) {
            *res = ans;
        }
        return;
    }

    int last_max = (N == 1) ? 1 : A[N - 2];
    for (int i = last_max; i <= M; i++) {
        A[N - 1] = i;
        rec(N - 1, M, Q, a, b, c, d, A, res);
    }
}

void solve(int N, int M, int Q, int *a, int *b, int *c, int *d) {
    int res = 0;
    int *A = (int *)malloc(N * sizeof(int));
    rec(N, M, Q, a, b, c, d, A, &res);
    printf("%d\n", res);
    free(A);
}

int main() {
    int N, M, Q;
    scanf("%d %d %d", &N, &M, &Q);
    int *a = (int *)malloc(Q * sizeof(int));
    int *b = (int *)malloc(Q * sizeof(int));
    int *c = (int *)malloc(Q * sizeof(int));
    int *d = (int *)malloc(Q * sizeof(int));
    for (int i = 0; i < Q; i++) {
        scanf("%d %d %d %d", &a[i], &b[i], &c[i], &d[i]);
    }
    solve(N, M, Q, a, b, c, d);
    free(a);
    free(b);
    free(c);
    free(d);
    return 0;
}

// End of Code.
